source/AST/ASTVisitor.cpp

```diff
@@ -622,7 +622,92 @@ shouldExtract(
 
     IsFileInRootDir_ = true;
 
-    return true;
+    if (config_.excludeNamespaces.empty())
+        return true;
+
+    const Decl * scope = nullptr;
+
+    const auto isScope (D->getKind() == Decl::Namespace ||  D->getKind() == Decl::CXXRecord);
+    if (isScope)
+        scope = D;
+    else
+        scope = dyn_cast<Decl>(D->getDeclContext());
+
+    return !calculateScopeFilter(scope);
+}
+
+bool
+ASTVisitor::
+calculateScopeFilter(const Decl * d)
+{
+    // Step 1: Check if we're talking about a scope type
+    if ((d == nullptr) ||
+        (d->getKind() != Decl::Namespace
+      && d->getKind() != Decl::Record))
+        return false;
+
+    auto itr = scopeFilter_.find(d);
+    if (itr != scopeFilter_.end())
+        return itr->second;
+
+    // check if any parent is skipped already
```

Comment:
I suspect this might not be needed, if we are in fact traversing top down as we should be.

Reply:
Is that true if we renter a namespace later?

Reply:
how can we re-enter a namespace?

Reply:
@sdkrystian can you please verify this